archive/logic_puzzles/knights_and_knaves.lean

```diff
@@ -0,0 +1,82 @@
+/-
+Copyright (c) 2020 Dan Stanescu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: D. Stanescu.
+-/
+
+import tactic
+
+/-!
+# Knights and knaves puzzles
+Two puzzles from "Knights and Knaves" by Raymond Smullyan.
+
+There is an island where all inhabitants are either knights or knaves. 
+Knights always tell the truth. Knaves always lie.
+Jack and Bob are two inhabitants of this island.
+-/
+
+inductive person
+| knight
+| knave
+
+notation `y` := person.knight
+notation `n` := person.knave
+
+structure Q := (p₁ p₂ : person)
+
+/--
+In this first puzzle, only Jack makes a statement.
+He affirms that both our heroes (i.e. both Jack and Bob) are knaves.
+What kind of islanders are they?
+-/
+
+def Q1 := Q
+
+namespace Q1
+
+variables (q : Q1)
+
+def S1 := q.1 = n ∧ q.2 = n
+
+def H := (q.1 = y ∧ q.S1) ∨ (q.1 = n ∧ ¬ q.S1)
+
+lemma answer : q.H → q.1 = n ∧ q.2 = y :=
+begin
+    rcases q with ⟨_|_,_|_⟩;
+    { simp [H, S1], },
+    done
+end
+
+end Q1
+
+/--
+In the second puzzle, both Jack and Bob make a statement.
+Jack states that he's a knave if and only if Bob is a knave.
+Bob only states that they are different kinds. 
+Again, what kind of islanders are our protagonists?
+-/
+
+
+def Q2 := Q
+
+namespace Q2
+
+variables (q : Q2)
+
+def S1 := q.2 = n ↔ q.1 = n
+def S2 := q.1 = y ∧ q.2 = n ∨ q.1 = n ∧ q.2 = y
```

Comment:
Perhaps truer to the statement would be `q.1 \ne q.2`, which then requires slightly more work in the proof!

(Maybe this is the wrong place for it, but it's perhaps never too early to realise that how you formalise a statement changes how you prove it!)

Reply:
Since they're both the same, I thought writing it this way shows exactly that there are different ways to formulate the problem. Some have advantages over the others. I think I could actually add another formulation of the same problem to the file.

Reply:
I think that would be best. Either stick exactly to the informal statement, or explicitly say that you're not (ideally by showing how things work out differently with the two different formulations).

Reply:
OK. The proof doesn't change but I'll add both.